lib/src/optimism/mod.rs

```diff
@@ -0,0 +1,348 @@
+            bytes32 batcher_hash,
+            uint256 l1_fee_overhead,
+            uint256 l1_fee_scalar
+        );
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct DeriveInput<D> {
+    pub db: D,
+    pub op_head_block_no: u64,
+    pub op_derive_block_count: u64,
+}
+
+#[derive(Debug, Clone, Deserialize, Eq, PartialEq, Serialize)]
+pub struct DeriveOutput {
+    pub eth_tail: (BlockNumber, BlockHash),
+    pub op_head: (BlockNumber, BlockHash),
+    pub derived_op_blocks: Vec<(BlockNumber, BlockHash)>,
+}
+
+pub struct DeriveMachine<D> {
+    pub derive_input: DeriveInput<D>,
+    op_head_block_hash: BlockHash,
+    op_block_no: u64,
+    op_block_seq_no: u64,
+    op_batcher: Batcher,
+    eth_block_no: u64,
+}
+
+impl<D: BatcherDb> DeriveMachine<D> {
+    pub fn new(mut derive_input: DeriveInput<D>) -> Result<Self> {
+        let op_block_no = derive_input.op_head_block_no;
+
+        // read system config from op_head (seq_no/epoch_no..etc)
+        let op_head = derive_input.db.get_full_op_block(op_block_no)?;
+        let op_head_block_hash = op_head.block_header.hash();
+
+        #[cfg(not(target_os = "zkvm"))]
+        info!(
+            "Fetched Op head (block no {}) {}",
+            derive_input.op_head_block_no, op_head_block_hash
+        );
+
+        let set_l1_block_values = {
+            let system_tx_data = op_head
+                .transactions
+                .first()
+                .unwrap()
+                .essence
+                .data()
+                .to_vec();
+            let call = OpSystemInfo::OpSystemInfoCalls::abi_decode(&system_tx_data, true)
+                .expect("Could not decode call data");
+            match call {
+                OpSystemInfo::OpSystemInfoCalls::setL1BlockValues(x) => x,
+            }
+        };
+
+        let op_block_seq_no = set_l1_block_values.sequence_number;
+
+        let eth_block_no = set_l1_block_values.number;
+        let eth_head = derive_input.db.get_eth_block_header(eth_block_no)?;
+        let eth_head_hash = eth_head.hash();
+        if eth_head_hash != set_l1_block_values.hash.as_slice() {
+            bail!("Ethereum head block hash mismatch.")
+        }
+        #[cfg(not(target_os = "zkvm"))]
+        info!(
+            "Fetched Eth head (block no {}) {}",
+            eth_block_no, set_l1_block_values.hash
+        );
+
+        let op_batcher = {
+            let mut op_chain_config = ChainConfig::optimism();
+            op_chain_config.system_config.batch_sender =
+                Address::from_slice(&set_l1_block_values.batcher_hash.as_slice()[12..]);
+            op_chain_config.system_config.l1_fee_overhead = set_l1_block_values.l1_fee_overhead;
+            op_chain_config.system_config.l1_fee_scalar = set_l1_block_values.l1_fee_scalar;
+
+            Batcher::new(
+                op_chain_config,
+                State::new(
+                    eth_block_no,
+                    eth_head_hash,
```

Comment:
Yeah, this part is strange. Good catch. The initialization logic needs some work. I will revisit this.

Reply:
Ok, I have reworked this logic. The new behavior is:

`Batcher::new()` receives the eth block corresponding to the Op head's attributes. It processes that block looking for deposits. It does not attempt to read any batches from it, though. (We know that it cannot contain any relevant batcher transactions. From the spec: *The L1 origin will never contain any data needed to construct sequencer batches since each batch must contain the L1 origin hash.*)

`DeriveMachine` does not invoke `Batcher::process_l1_block()` on its first iteration. This is because the first Eth block has effectively been processed already by `Batcher::new()`.

With this change, `DeriveMachine` no longer needs to keep track of the next Eth block to fetch. Instead, it can rely on the state held within `Batcher`.

Lastly, with this change, the `Batcher` state now indicates the last L1 block that was processed. (Previously its semantics were less clear.)

Reply:
Marking this thread as resolved, since it now refers to an older commit.